dffml/util/net.py

```diff
@@ -195,12 +211,26 @@ def cached_download_unpack_archive(
     >>> asyncio.run(files_in_dffml_commit_c4469a())
     124
     """
+
+    def on_error(func, path, exc_info):
+        """
+        If there's permission denied error. Error handler for shutil.rmtree
+        """
+        if not os.access(path, os.W_OK):
+            os.chmod(path, stat.S_IWUSR)
+            func(path)
+        else:
+            raise DirectoryNotExtractedError(path)
+
     directory_path = pathlib.Path(directory_path)
 
     async def extractor(download_path):
         download_path = download_path.absolute()
         with chdir(directory_path):
-            shutil.unpack_archive(str(download_path), ".")
+            try:
+                shutil.unpack_archive(str(download_path), ".")
+            except:
+                shutil.rmtree(directory_path, onerror=on_error)
```

Comment:
- I don't think we're using `onerror` in quite the way we should here. `rmtree`'s `onerror` is for when there are issues removing the extracted file.
  - We want to try to unpack the archive, if we fail, then we remote the directory, then we raise the error that occurred when unpacking. Or else the user won't know what happened. They'll be confused as to why we were trying and failing to delete their partially extracted archive.
  - It would be good to `except Exception as error`, then you can `raise DirectoryNotExtractedError(path) from error`.

- Can we get a test for this case? `unittest.mock` will be helpful here. There are examples throughout the tests (`git grep unittest.mock` or `git grep with\ patch`). You should be able to make `unpack_archive` raise an exception.

Reply:
Actually! in windows, the common problem with rmtree is that it throws an exception of permission denied. So, to handle this I have added the onerror. But yeah! you are right. User may be confused about what is going on.
Now, I made the changes accordingly so that whenever we failed to unpack it raise an exception ```DirectoryNotExtractedError```.